tstest/clock.go

```diff
@@ -4,57 +4,495 @@
+	return cb
+}
+
+func (cb *ClockBuilder) Step(d time.Duration) *ClockBuilder {
+	cb.c.step = d
+	return cb
+}
+
+func (cb *ClockBuilder) TimerChannelSize(n int) *ClockBuilder {
+	cb.c.timerChannelSize = n
+	return cb
+}
+
+func (cb *ClockBuilder) Finish() *Clock {
+	c := cb.c
+	cb.c = nil
+	c.init()
+	return c
+}
+
+// NewClock returns a Clock initialized with default settings (equivalent to
+// NewClockBuilder().Finish()).
+func NewClock() *Clock {
+	c := new(Clock)
+	c.init()
+	return c
+}
+
 // Clock is a testing clock that advances every time its Now method is
-// called, beginning at Start.
-//
-// The zero value starts virtual time at an arbitrary value recorded
-// in Start on the first call to Now, and time never advances.
+// called, beginning at its start time. If no start time is specified using
+// ClockBuilder, an arbitrary start time will be selected when the Clock is
+// created and can be retrieved by calling Clock.Start().
 type Clock struct {
-	// Start is the first value returned by Now.
-	Start time.Time
-	// Step is how much to advance with each Now call.
-	Step time.Duration
-	// Present is the time that the next Now call will receive.
-	Present time.Time
+	// start is the first value returned by Now. It must not be modified after
+	// init() is called.
+	start time.Time
+
+	mu sync.Mutex
 
-	sync.Mutex
+	// step is how much to advance with each Now call.
+	step time.Duration
+	// present is the last value returned by Now (and will be returned again by
+	// ConstantNow).
+	present time.Time
+	// skipStep indicates that the next call to Now should not add step to
+	// present. This occurs after initialization and after Advance.
+	skipStep bool
+	// timerChannelSize is the buffer size to use for channels created by
+	// NewTimer and NewTicker.
+	timerChannelSize int
+
+	events eventManager
+}
+
+func (c *Clock) init() {
+	if c.start.IsZero() {
+		c.start = time.Now()
+	}
+	if c.timerChannelSize == 0 {
+		c.timerChannelSize = 1
+	}
+	c.present = c.start
+	c.skipStep = true
+	c.events.AdvanceTo(c.present)
 }
 
 // Now returns the virtual clock's current time, and advances it
 // according to its step configuration.
 func (c *Clock) Now() time.Time {
-	c.Lock()
-	defer c.Unlock()
-	c.initLocked()
-	step := c.Step
-	ret := c.Present
-	c.Present = c.Present.Add(step)
-	return ret
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.skipStep && c.step > 0 {
+		c.present = c.present.Add(c.step)
+		c.events.AdvanceTo(c.present)
+	}
+	c.skipStep = false
+	return c.present
+}
+
+// ConstantNow returns the virtual clock's current time but does not advance it
+// if a step is configured.
+func (c *Clock) ConstantNow() time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.present
+}
+
+func (c *Clock) Advance(d time.Duration) time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.skipStep = true
+
+	if d > 0 {
+		c.present = c.present.Add(d)
+		c.events.AdvanceTo(c.present)
+	}
+
+	return c.present
+}
+
+func (c *Clock) Start() time.Time {
```

Comment:
Hmm, Start() sounds like a command to start the clock running, rather than an accessor. Maybe StartTime() ? Or maybe with a docstring and then used in context it doesn't matter. Not a strong opinion.

Reply:
I feel like StartTime() risks being stuttery? I made it `GetStart` to be consistent with the mutation accessors (e.g. `SetStep`).